server/src/test/java/org/elasticsearch/common/collect/IteratorsTests.java

```diff
@@ -110,6 +111,37 @@ public void testNullIterator() {
         }
     }
 
+    public void testArrayIterator() {
+        Integer[] array = randomIntegerArray();
+        Iterator<Integer> iterator = Iterators.forArray(array);
+
+        int i = 0;
+        while (iterator.hasNext()) {
+            assertEquals(array[i++], iterator.next());
+        }
+    }
+
+    public void testArrayIteratorIsUnmodifiable() {
+        Integer[] array = randomIntegerArray();
+        Iterator<Integer> iterator = Iterators.forArray(array);
+
+        expectThrows(UnsupportedOperationException.class, iterator::remove);
+    }
+
+    public void testArrayIteratorThrowsNoSuchElementExceptionWhenDepleted() {
+        Integer[] array = randomIntegerArray();
+        Iterator<Integer> iterator = Iterators.forArray(array);
+        for (int i = 0; i < array.length; i++) {
+            iterator.next();
+        }
+
+        expectThrows(NoSuchElementException.class, iterator::next);
+    }
+
+    private static Integer[] randomIntegerArray() {
+        return Randomness.get().ints(randomIntBetween(1, 1000)).boxed().toArray(Integer[]::new);
```

Comment:
I think it's important also to verify that it behaves correctly with empty arrays:

```suggestion
        return Randomness.get().ints(randomIntBetween(0, 1000)).boxed().toArray(Integer[]::new);
```